8293862: javax/swing/JFileChooser/8046391/bug8046391.java failed with 'Cannot invoke "java.awt.Image.getWidth(java.awt.image.ImageObserver)" because "retVal" is null'

TejeshR13 opened the pull request:
Observation found when JFileChooser is instantiated in WindowsLookAndFeel which invokes getSystemIcon() from WindowsFileChooserUI class. Could not find the exact root cause so predicting it to be an issue with icons not loaded where resolutionVariants map is empty in _public Image getResolutionVariant(double width, double height) _. Hence proposing a null check before accessing getWidth(). Fix is tested in CI system.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Error
&nbsp;⚠️ 8293862 is used in problem lists: [test/jdk/ProblemList.txt]

### Issue
 * [JDK-8293862](https://bugs.openjdk.org/browse/JDK-8293862): javax/swing/JFileChooser/8046391/bug8046391.java failed with 'Cannot invoke "java.awt.Image.getWidth(java.awt.image.ImageObserver)" because "retVal" is null'


### Reviewers
 * [Abhishek Kumar](https://openjdk.org/census#abhiscxk) (@kumarabhi006 - Author)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11104/head:pull/11104` \
`$ git checkout pull/11104`

Update a local copy of the PR: \
`$ git checkout pull/11104` \
`$ git pull https://git.openjdk.org/jdk pull/11104/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11104`

View PR using the GUI difftool: \
`$ git pr show -t 11104`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11104.diff">https://git.openjdk.org/jdk/pull/11104.diff</a>

</details>

aivanov-jdk commented:
Can you verify that the MultiResolutionImage contains no variants?

If it's the case, a cleaner way would be to return `null` right away.

Or rather, the function that creates the `MultiResolutionIconImage` instance should be modified to return `null` rather than returning an MRI with no variants. This should never happen because if the system does not provide an icon, a default icon gets requested.

By adding the null-check we don't resolve the real problem but pretend it does not exist.

prrace commented:
I agree with Alexey.

TejeshR13 commented:
> 

Yeah, the only way the retVal can become null would-be if `resolutionVariants` doesn't have an Icon. I could not reproduce the bug since it is intermittent, I came to this conclusion only by code analysis though. Should I check for `resolutionVariants` if its empty and return null without proceeding further......? I guess both the ways `null` will be returned?

aivanov-jdk commented:
> Yeah, the only way the retVal can become null would-be if `resolutionVariants` doesn't have an Icon. I could not reproduce the bug since it is intermittent, I came to this conclusion only by code analysis though. Should I check for `resolutionVariants` if its empty and return null without proceeding further......? I guess both the ways `null` will be returned?

An MRI that has no images does not make any sense, we should've returned null.

There are two usages of the constructor which accepts resolution variants:
https://github.com/openjdk/jdk/blob/6c8d0e617ff59eee1313589b10edbf5830774db5/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L1185
https://github.com/openjdk/jdk/blob/6c8d0e617ff59eee1313589b10edbf5830774db5/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L1112-L1113

There are four usages of the other constructor which accepts an image.

Since it's an intermittent problem, we need to add debugging code to the code which creates the `MultiResolutionIconImage` object or handle the exceptional cases in the constructor… by adding an `assert` statement. The tests are run with asserts enabled and we may get more data.

Is there a host where the test is more likely to fail? Does it fail on different hosts?

Adding the null check to `getResolutionVariant` does not fix the root cause. Instead, it'll make it impossible to find the root cause because everything would look as if no problem exists.

TejeshR13 commented:
> > Yeah, the only way the retVal can become null would-be if `resolutionVariants` doesn't have an Icon. I could not reproduce the bug since it is intermittent, I came to this conclusion only by code analysis though. Should I check for `resolutionVariants` if its empty and return null without proceeding further......? I guess both the ways `null` will be returned?
> 
> An MRI that has no images does not make any sense, we should've returned null.
> 
> There are two usages of the constructor which accepts resolution variants:
> 
> https://github.com/openjdk/jdk/blob/6c8d0e617ff59eee1313589b10edbf5830774db5/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L1185
> 
> 
> https://github.com/openjdk/jdk/blob/6c8d0e617ff59eee1313589b10edbf5830774db5/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L1112-L1113
> 
> There are four usages of the other constructor which accepts an image.
> 
> Since it's an intermittent problem, we need to add debugging code to the code which creates the `MultiResolutionIconImage` object or handle the exceptional cases in the constructor… by adding an `assert` statement. The tests are run with asserts enabled and we may get more data.
> 
> Is there a host where the test is more likely to fail? Does it fail on different hosts?
> 
> Adding the null check to `getResolutionVariant` does not fix the root cause. Instead, it'll make it impossible to find the root cause because everything would look as if no problem exists.

Yeah, agree to your point that it actually mask the actual problem. It fails only when windows L&F is used and `JFileChooser` is instantiated. On creating file chooser instance on windows `getSystemIcon` is called initiates the problem.

dcubed-ojdk commented:
This PR hasn't been touched in 8 days. I'm going to ProblemListing this test
on windowx-x64 since we're up to 7 Tier3 failures in the jdk/jdk CI.

aivanov-jdk commented:
Tejesh is testing the approach to get more details about the failure.

I agree the issue is creating noise.

If you problem-list the test (the two tests), he will be able to remove the test(s) from the problem list for his experiments.

TejeshR13 commented:
> This PR hasn't been touched in 8 days. I'm going to ProblemListing this test on windowx-x64 since we're up to 7 Tier3 failures in the jdk/jdk CI.

Working on getting more info about the failure since the issue is intermittent. I am testing the diagnostic fix such that it won't affect the other test.

prrace commented:
Where are we at with making a diagnostic update to this code and the failing test(s) ?

TejeshR13 commented:
> Where are we at with making a diagnostic update to this code and the failing test(s) ?

It looks like the imageIcon were not loaded for _downloads_ folder, which might be the cause for `retVal` getting null value instead of ImageIcon. Mach5 tests link attached in JBS.

aivanov-jdk commented:
> > Where are we at with making a diagnostic update to this code and the failing test(s) ?
> 
> It looks like the imageIcon were not loaded for _downloads_ folder, which might be the cause for `retVal` getting null value instead of ImageIcon. Mach5 tests link attached in JBS.

Is there a reason why getting the icon for _Downloads_ folder fails?

TejeshR13 commented:
@dcubed-ojdk I'm able to reproduce the bug in mach5 test runs more often. More of failures are observed within test runs also and the issue is with `downloads` ImageIcon not loading up, to find out the exact root cause I might need some more time. Problem listing the test would be better now.

prsadhuk commented:
Do we need to wait a bit longer after creating JFileChooser as the test immediately ends after instantiating it?
Also, doesn't JFileChooser needs headful keyword as it needs to show a file dialog? Does it work more reliably if its to run in headful CI systems?

TejeshR13 commented:
> Do we need to wait a bit longer after creating JFileChooser as the test immediately ends after instantiating it? Also, doesn't JFileChooser needs headful keyword as it needs to show a file dialog? Does it work more reliably if its to run in headful CI systems?

Should the test be headful even if we instantiate JFileChooser without showing the dialog.? I can check with a delay after its instantiated, but no sure if that might be causing the issue. Because the same issue is observed in another bug also which instantiates JFileChooser and does some tasks after that.

aivanov-jdk commented:
> > Do we need to wait a bit longer after creating JFileChooser as the test immediately ends after instantiating it? Also, doesn't JFileChooser needs headful keyword as it needs to show a file dialog? Does it work more reliably if its to run in headful CI systems?
> 
> Should the test be headful even if we instantiate JFileChooser without showing the dialog.? I can check with a delay after its instantiated, but no sure if that might be causing the issue. Because the same issue is observed in another bug also which instantiates JFileChooser and does some tasks after that.

It doesn't need to be. It would've failed if it had required a headful system.

However, it may still be a valid concern. I faced some obscure failures on headless systems running Windows.

The exception is thrown on EDT when `JFileChooser` calculates its preferred size, so the test seem to work correctly. As such, I don't think the failure is related to the fact that the test exits after creating `JFileChooser`.

One of the options to fix the test is to add `@key headful`.

I wonder if the _Downloads_ folder is the first folder for which the icon is requested. Does fetching icons for other folders also fail? Does the host where the test fails have Windows Explorer? Windows Server systems can have minimal UI, if Windows Explorer is not available, then the `headful` keyword is the way to go.

So, we need to collect more data why it fails, why the icon of a folder is `null`.

In addition to that, we have to ensure `null` is returned directly to the caller rather than creating a `MultiResolutionImage` which doesn't even contain a base image.

mrserb commented:
I suggest calling the FileSystemView API directly on that folder, I think that should work in headless and headfull systems.

TejeshR13 commented:
> I suggest calling the FileSystemView API directly on that folder, I think that should work in headless and headfull systems.
@mrserb I didn't get you. Can you brief a little more w.r.t "calling the FileSystemView API"..?

mrserb commented:
JFileChooser uses the FileSystemView API, you can call that directly to make sure that some UI things in JFileChooser do not affect the behavior of the test.

TejeshR13 commented:
> JFileChooser uses the FileSystemView API, you can call that directly to make sure that some UI things in JFileChooser do not affect the behavior of the test.

The test only initializes JFileChooser instance, so calling FSV API directly means to explicitly mentioning FSV from test or from JFileChooser...?

mrserb commented:
> The test only initializes JFileChooser instance, so calling FSV API directly means to explicitly mentioning FSV from test or from JFileChooser...?

No, just use that for debugging.

TejeshR13 commented:
> > The test only initializes JFileChooser instance, so calling FSV API directly means to explicitly mentioning FSV from test or from JFileChooser...?
> 
> No, just use that for debugging.

Ok, will try that, thank you for the pointers.

TejeshR13 commented:
@aivanov-jdk I just run some test samples by making the test as headful instead as headless as @prsadhuk suggested. I found out that making it as headful works just fine and no NPE occurred (Headless always showed issue, 8/200 test). And debugging to the root cause of Image not loading is also quite difficult. I guess it's better to modify the test as headful since the reliability of the test is better in it?

aivanov-jdk commented:
> @aivanov-jdk I just run some test samples by making the test as headful instead as headless as @prsadhuk suggested. I found out that making it as headful works just fine and no NPE occurred (Headless always showed issue, 8/200 test). And debugging to the root cause of Image not loading is also quite difficult. I guess it's better to modify the test as headful since the reliability of the test is better in it?

Let's go for it. A headless system could lack something that the test uses even though it doesn't show the UI.

If _headful_ systems never reproduce the problem, then it's _the fix_.

aivanov-jdk commented:
At the same time, I suggest filing another bug to track the issue where the code could return a `MultiResolutionImage` with no base image. It should never happen.

Or you can address it here along with making the test _headful_.

TejeshR13 commented:
> 

Means checking for null when trying to insert null into MRI map?

TejeshR13 commented:
> > @aivanov-jdk I just run some test samples by making the test as headful instead as headless as @prsadhuk suggested. I found out that making it as headful works just fine and no NPE occurred (Headless always showed issue, 8/200 test). And debugging to the root cause of Image not loading is also quite difficult. I guess it's better to modify the test as headful since the reliability of the test is better in it?
> 
> Let's go for it. A headless system could lack something that the test uses even though it doesn't show the UI.
> 
> If _headful_ systems never reproduce the problem, then it's _the fix_.

So far, from the analysis it is found that the backup mechanism to get default Icon is failing when icon extraction fails. Not able to get the trigger point for the failure since it is unpredictable. Failure rate also seems to be dependent on system, last week used to get consistent failures for every test run, this week the issue is not much. So I guess making the test headful is better option now.

aivanov-jdk commented:
> Means checking for null when trying to insert null into MRI map?

Yes, because such an image is invalid — it will produce the NPE when it's rendered. If icon isn't available, `null` is to be returned directly to the caller.

TejeshR13 commented:
> > So far, from the analysis it is found that the backup mechanism to get default Icon is failing when icon extraction fails.
> 
> These are the new details which you've never mentioned before.
> 
> The icon extraction of a folder (presumably _Downloads_ folder) fails. The fallback is to extract the default icon for a folder, which also fails.
> 
> Do I understand it correctly?

Yes. And the fall back is failing for small icon which has size 16. Last week issue was reproducible very frequently, but today didn't get any failure for test runs.

TejeshR13 commented:
Based on the analysis and test runs, I have added a null check to return null when _ImageIcon_ is null and added to the map. The test is made as headful since the occurrence of issue seems to be dependent on machine and is unpredictable. Headful test didn't cause any issue after multiple runs and iterations.

TejeshR13 commented:
@aivanov-jdk I have updated the PR with the above fix, can you please review and let me know if it's Ok.

aivanov-jdk commented on src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java:
```diff
@@ -1437,7 +1441,7 @@ public Image getResolutionVariant(double width, double height) {
                     }
                 }
             }
-            if (retVal.getWidth(null) != w) {
+            if ((retVal != null) && (retVal.getWidth(null) != w)) {
```
I believe this change is not needed.

TejeshR13 replied:
> You may still add `assert` statements into `MultiResolutionIconImage` constructors which ensures `resolutionVariants` contains no `null` values.

Will add assert statements in this `public MultiResolutionIconImage(int baseSize, Image image)` as null checks are handled in other constructor.

aivanov-jdk replied:
> > You may still add `assert` statements into `MultiResolutionIconImage` constructors which ensures `resolutionVariants` contains no `null` values.
> 
> Will add assert statements in this `public MultiResolutionIconImage(int baseSize, Image image)` as null checks are handled in other constructor.

What do you mean?

There are no `null`-checks in the other constructor. Or do I miss anything?

It's just you don't call the other constructor any more if an icon fetched from the OS is `null`. (Perhaps, it's not the most effective way: if another icon size is extracted correctly, a non-null value could still be returned.)

The following statement

```java
assert !resolutionVariants.containsValue(null) : "There are null icons in the variants map";
```

would still prevent a `MultiResolutionIconImage` from being created if any resolution variant contains `null`. The code above could change again and we may end up with the same problem again.

TejeshR13 replied:
Means the other constructor is called in two places only and in those places null checks are done. So, even if we add assertion inside constructor, it will unreachable.

aivanov-jdk replied:
> Means the other constructor is called in two places only and in those places null checks are done. So, even if we add assertion inside constructor, it will unreachable.

It won't be unreachable, _it will prevent such a situation from happening again_. The code which calls the other constructor could be changed.

I still think it's reasonable to add the assertion. It's a no-op if assertions aren't enabled (the default), and it'll catch the broken invariant in testing since tests are (usually) run with assertions enabled.

TejeshR13 replied:
Yeah, got your point. It's better to add as a preventive measure.

aivanov-jdk reviewed (changes requested):
You may still add `assert` statements into `MultiResolutionIconImage` constructors which ensures `resolutionVariants` contains no `null` values.

aivanov-jdk commented on src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java:
```diff
@@ -1399,11 +1399,13 @@ static class MultiResolutionIconImage extends AbstractMultiResolutionImage {
         }
 
         public MultiResolutionIconImage(int baseSize, Image image) {
+            assert image != null : "Null icons are added into MRI variants map";
```
```suggestion
            assert image != null : "Null icon passed as the base image for MRI";
```

TejeshR13 replied:
Updated.

aivanov-jdk reviewed (approved):
Approved provided the message for MRII(int, Image) is updated.

mrserb commented:
Did we check how the native code works w/o using JFileChooser, buy FileSystemView, or by the native example? Is it possible that it is intermittent because of timing related to "com" thread?

TejeshR13 commented:
> Did we check how the native code works w/o using JFileChooser, buy FileSystemView, or by the native example? Is it possible that it is intermittent because of timing related to "com" thread?



> 

I have tested in headful system multiple time, never reproduced this issue...... Can timing affect headless......?

mrserb commented:
> I have tested in headful system multiple time, never reproduced this issue...... Can timing affect in headless and not in headful......?

It should not. Does it actually have any dependency on "headful" envirement? The native code we used does not create any UI element or something like that. It is better to debug it on the same/similar config/system where the bug is reproduced, since as far as I understand it is not a rare case.

TejeshR13 commented:
> > I have tested in headful system multiple time, never reproduced this issue...... Can timing affect in headless and not in headful......?
> 
> It should not. Does it actually have any dependency on "headful" envirement? The native code we used does not create any UI element or something like that. It is better to debug it on the same/similar config/system where the bug is reproduced, since as far as I understand it is not a rare case.

I don't think there is any dependency on "headful". The occurrence was very random, like sometime issue used to show up repeatedly for test runs, and the very next week there were no occurrences, and the occurrence was not fixed for any particular path/folder, it used to vary (For headful). I was able to debug and understand that backup mechanism failed at random intervals, so had to conclude that it had to something to do with headful.

aivanov-jdk commented:
@mrserb, @TejeshR13, There could be a dependency on the host being headful even if we don't create UI. You may remember there were a couple of test failures where colours in `VolatileImage` were different from the colours in `BufferedImage` because `VolatileImage` used `IndexedColorModel` rather than the expected 32-bit colour.

Recently I saw another problem where a seemingly headless test didn't work as expected on the headless system. Windows L&F wasn't set correctly and `JTextField` didn't have the expected border.

Windows Server systems may not have full-featured desktop environment, in this case we may see problems with *headless* tests which in fact need *headful (full-featured)* desktop environment even though such tests don't show the UI.

I agree it is worth exploring further why fetching the icon of the _Downloads_ folder returns `null`. Could either of you submit a bug, please?

TejeshR13 commented:
> @mrserb, @TejeshR13, There could be a dependency on the host being headful even if we don't create UI. You may remember there were a couple of test failures where colours in `VolatileImage` were different from the colours in `BufferedImage` because `VolatileImage` used `IndexedColorModel` rather than the expected 32-bit colour.
> 
> Recently I saw another problem where a seemingly headless test didn't work as expected on the headless system. Windows L&F wasn't set correctly and `JTextField` didn't have the expected border.
> 
> Windows Server systems may not have full-featured desktop environment, in this case we may see problems with _headless_ tests which in fact need _headful (full-featured)_ desktop environment even though such tests don't show the UI.
> 
> I agree it is worth exploring further why fetching the icon of the _Downloads_ folder returns `null`. Could either of you submit a bug, please?

Not only _Downloads_ folder returns null, as I was mentioning it was varying for multiple test runs. Then submitting a new bug for ImageIcon failing to load Image in headless systems right?

TejeshR13 commented:
> > Not only _Downloads_ folder returns null, as I was mentioning it was varying for multiple test runs.
> 
> I remember you were saying that only this one was affected. I may have missed that other folders were affected as well.
> 
> > Then submitting a new bug for ImageIcon failing to load Image in headless systems right?
> 
> Yes, for this particular problem: a folder icon cannot be fetched from Windows on a headless system.

Created - [JDK-8299893](https://bugs.openjdk.org/browse/JDK-8299893). Shall I integrate this PR by removing the test from problemList?

mrserb commented:
> Created - [JDK-8299893](https://bugs.openjdk.org/browse/JDK-8299893). Shall I integrate this PR by removing the test from problemList?

Did we have a chance to reproduce it at least once other than accidentally in mach5? If not that not necessarily is a root cause, I doubt that the new bug will not be fixed any time soon.

mrserb commented on test/jdk/javax/swing/JFileChooser/8046391/bug8046391.java:
```diff
@@ -27,6 +27,7 @@
  * @requires (os.family == "windows")
  * @summary JFileChooser hangs if displayed in Windows L&F
  * @author Alexey Ivanov
+ * @key headful
```
If we assume that the bug is fixed by the code change we do not need to change the test.

TejeshR13 replied:
Since we are not 100% sure that the issue is not with headless and is with code, I guess it is safe to change the test to headful. Anyhow it will take time to analyse and conclude that the issue is with the code and so the new bug has been raised. This PR addresses handling the error cases also if you can go through the previous conversations, so the new can bug can be focused on whether the bug is w.r.t headless/code.

TejeshR13 commented:
> > Created - [JDK-8299893](https://bugs.openjdk.org/browse/JDK-8299893). Shall I integrate this PR by removing the test from problemList?
> 
> Did we have a chance to reproduce it at least once other than accidentally in mach5? If not that not necessarily is a root cause, I doubt that the new bug will not be fixed any time soon.

Other than accidentally in mach5, NO. Out of 200 repeated tests, 8 tests failures were observed, again the numbers are not fixed.

aivanov-jdk commented:
@mrserb was absolutely right! Marking the test `@headful` will not resolve the problem, it will make us forget about the problem, and apparently it may manifest itself on headful systems too.

I added my own traces and I've found three issues with the code.

**`E_PENDING`**
https://github.com/openjdk/jdk/blob/8cb25d3de494c6d9357a1c199e1a9dbff9be9948/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L1157-L1158

An icon handle could be negative: `hIcon = 0xffffffffa25400f7` is perfectly valid handle but it's rejected by our current code.

The same problem exists two lines below:
https://github.com/openjdk/jdk/blob/8cb25d3de494c6d9357a1c199e1a9dbff9be9948/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L1161

**Result of `Extract`**
https://github.com/openjdk/jdk/blob/8cb25d3de494c6d9357a1c199e1a9dbff9be9948/src/java.desktop/windows/native/libawt/windows/ShellFolder2.cpp#L997

The result of calling isn't verified.

This second error led to returning an invalid handle to the *(default)* folder icon. I saw it only for 16×16 icons, in all these cases something that looks like handle is returned, that is `hIcon` isn't null. Yet when the icon bits are extracted, `fn_GetIconInfo` returns an error code 0x57a - _Invalid cursor handle_.

The two errors combined lead to quite a few *random* failures. No particular host reproduced the problem more often. Sometimes 5 of 10 tests failed, sometimes only 1 of 10. Yet none of the test runs I submitted succeeded before I resolved the above issues.

I'll re-assign the bug to myself and submit a new PR after I do additional tests.

aivanov-jdk reviewed (changes requested):
Review is revoked as per my recent comment.

aivanov-jdk commented:
This PR is superseded by #12010.

prrace commented:
This PR should be withdrawn
